File: 2020/python/day06.py
#!/bin/env python3

from typing import List, Tuple, Dict

DAY = "day6"
INPUT = "input/2020/" + DAY + ".txt"


def main():
    groups = breakdown()
    print("--- Part 1 ---\n")
    print("Sum: " + str(sum_len(groups)))
    print()
    print("--- Part 2 ---\n")
    print("Sum: " + str(num_same_answers(groups)))


def sum_len(groups: List[Tuple[int, Dict]]) -> int:
    return sum([len(group[1]) for group in groups])


def num_same_answers(groups: List[Tuple[int, Dict]]) -> int:
    right = 0
    for group in groups:
        right += len([x for x in group[1].values() if x == group[0]])
    return right


def breakdown() -> List[Tuple[int, Dict]]:
    groups = []
    with open(INPUT) as txt:
        content = txt.read().rstrip()
        for raw_groups in content.split("\n\n"):
            answers = {}
            people = 0
            for raw_group in raw_groups.split("\n"):
                people += 1
                for a in raw_group:
                    if a in answers:
                        answers[a] = answers[a] + 1
                    else:
                        answers[a] = 1
            groups.append((people, answers))
    return groups


if __name__ == "__main__":
    main()

File: 2020/go/day06.go
package aoc2020

import (
	"fmt"
	"strings"

	"github.com/elliotchance/pie/pie"
	"github.com/scottames/adventofcode/pkg/helpers"
)

const (
	day06 = 6
)

// Day06 - Day 6 Part 1 & 2
func Day06() error {
	input, err := helpers.ReadInput(year, day06)
	if err != nil {
		return err
	}

	groups := helpers.StringSplitNewlinesNewlinesStrings(input)

	// Part 1
	helpers.PrintPart1()
	fmt.Printf("Sum of counts: %d\n", sumCounts(groups))
	fmt.Println()

	// Part 2
	helpers.PrintPart2()
	fmt.Printf("Sum of counts: %d\n", sumCountsEveryone(groups))
	fmt.Println()

	return nil
}

// TODO: redo this solution in more efficient & cleaner way
func sumCounts(groups pie.Strings) int {
	answers := pie.Ints{}
	for _, group := range groups {
		trimmed := uniqueString(strings.Replace(group, "\n", "", -1))
		answers = answers.Append(len(trimmed))
	}
	return answers.Sum()
}

func sumCountsEveryone(groups pie.Strings) int {
	var y pie.Strings
	for _, group := range groups {
		var a map[rune]int = make(map[rune]int)
		people := strings.Split(group, "\n")
		for _, person := range people {
			for _, answer := range person {
				a[answer]++
			}
		}
		for answer, number := range a {
			if number == len(people) {
				y = y.Append(string(answer))
			}
		}
	}
	return y.Len()
}

func uniqueString(str string) map[string]int {
	result := map[string]int{}
	for _, c := range str {
		s := string(c)
		result[s]++

	}
	return result
}
